### ISSUE TITLE
Error MT1001: Could not find an application at the specified directory

### ISSUE THREAD

**Issue by munkii** [xamarin/xamarin-macios#5344]
### Steps to Reproduce

1. Successfully connect the specific project (**other projects work fine**) to the Mac Build Server
2. Build/Rebuild seem to both succeed
3. Deploy

### Expected Behavior
iOS app will deploy to either the Simulator or a Device
### Actual Behavior
Deploy fails "App terminated". Various errors appear in the Xamarin Logs which I have attached. No files seem to get created during build
### Environment

```
Microsoft Visual Studio Professional 2017 
Version 15.9.4
VisualStudio.15.Release/15.9.4+28307.222
Microsoft .NET Framework
Version 4.7.03056

Installed Version: Professional

Visual C++ 2017   00369-60000-00001-AA954
Microsoft Visual C++ 2017

Application Insights Tools for Visual Studio Package   8.14.11009.1
Application Insights Tools for Visual Studio

ASP.NET and Web Tools 2017   15.9.04012.0
ASP.NET and Web Tools 2017

ASP.NET Core Razor Language Services   15.8.31590
Provides languages services for ASP.NET Core Razor.

ASP.NET Web Frameworks and Tools 2017   5.2.60913.0
For additional information, visit https://www.asp.net/

Azure App Service Tools v3.0.0   15.9.03024.0
Azure App Service Tools v3.0.0

Azure Data Lake Node   1.0
This package contains the Data Lake integration nodes for Server Explorer.

Azure Data Lake Tools for Visual Studio   2.3.3000.2
Microsoft Azure Data Lake Tools for Visual Studio

Azure Functions and Web Jobs Tools   15.9.02046.0
Azure Functions and Web Jobs Tools

Azure Stream Analytics Tools for Visual Studio   2.3.3000.2
Microsoft Azure Stream Analytics Tools for Visual Studio

C# Tools   2.10.0-beta2-63501-03+b9fb1610c87cccc8ceb74a770dba261a58e39c4a
C# components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Common Azure Tools   1.10
Provides common services for use by Azure Mobile Services and Microsoft Azure Tools.

Extensibility Message Bus   1.1.49 (remotes/origin/d15-8@ee674f3)
Provides common messaging-based MEF services for loosely coupled Visual Studio extension components communication and integration.

Fabric.DiagnosticEvents   1.0
Fabric Diagnostic Events

JavaScript Language Service   2.0
JavaScript Language Service

JavaScript Project System   2.0
JavaScript Project System

JavaScript UWP Project System   2.0
JavaScript UWP Project System

JetBrains ReSharper Ultimate 2018.3.1   Build 183.0.20181225.200351
JetBrains ReSharper Ultimate package for Microsoft Visual Studio. For more information about ReSharper Ultimate, visit http://www.jetbrains.com/resharper. Copyright © 2019 JetBrains, Inc.

Microsoft Azure HDInsight Azure Node   2.3.3000.2
HDInsight Node under Azure Node

Microsoft Azure Hive Query Language Service   2.3.3000.2
Language service for Hive query

Microsoft Azure Service Fabric Tools for Visual Studio   2.4
Microsoft Azure Service Fabric Tools for Visual Studio

Microsoft Azure Stream Analytics Language Service   2.3.3000.2
Language service for Azure Stream Analytics

Microsoft Azure Stream Analytics Node   1.0
Azure Stream Analytics Node under Azure Node

Microsoft Azure Tools   2.9
Microsoft Azure Tools for Microsoft Visual Studio 2017 - v2.9.10730.2

Microsoft Continuous Delivery Tools for Visual Studio   0.4
Simplifying the configuration of Azure DevOps pipelines from within the Visual Studio IDE.

Microsoft JVM Debugger   1.0
Provides support for connecting the Visual Studio debugger to JDWP compatible Java Virtual Machines

Microsoft Library Manager   1.0
Install client-side libraries easily to any web project

Microsoft MI-Based Debugger   1.0
Provides support for connecting Visual Studio to MI compatible debuggers

Microsoft Visual C++ Wizards   1.0
Microsoft Visual C++ Wizards

Microsoft Visual Studio Tools for Containers   1.1
Develop, run, validate your ASP.NET Core applications in the target environment. F5 your application directly into a container with debugging, or CTRL + F5 to edit & refresh your app without having to rebuild the container.

Microsoft Visual Studio VC Package   1.0
Microsoft Visual Studio VC Package

MLGen Package Extension   1.0
MLGen Package Visual Studio Extension Detailed Info

Mono Debugging for Visual Studio   4.13.12-pre (9bc9548)
Support for debugging Mono processes with Visual Studio.

NuGet Package Manager   4.6.0
NuGet Package Manager in Visual Studio. For more information about NuGet, visit http://docs.nuget.org/.

ProjectServicesPackage Extension   1.0
ProjectServicesPackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

ResourcePackage Extension   1.0
ResourcePackage Visual Studio Extension Detailed Info

SQL Server Data Tools   15.1.61810.11040
Microsoft SQL Server Data Tools

ToolWindowHostedEditor   1.0
Hosting json editor into a tool window

TypeScript Tools   15.9.20918.2001
TypeScript Tools for Microsoft Visual Studio

Visual Basic Tools   2.10.0-beta2-63501-03+b9fb1610c87cccc8ceb74a770dba261a58e39c4a
Visual Basic components used in the IDE. Depending on your project type and settings, a different version of the compiler may be used.

Visual F# Tools 10.2 for F# 4.5   15.8.0.0.  Commit Hash: 6e26c5bacc8c4201e962f5bdde0a177f82f88691.
Microsoft Visual F# Tools 10.2 for F# 4.5

Visual Studio Code Debug Adapter Host Package   1.0
Interop layer for hosting Visual Studio Code debug adapters in Visual Studio

Visual Studio Tools for Containers   1.0
Visual Studio Tools for Containers

Visual Studio Tools for Universal Windows Apps   15.0.28307.208
The Visual Studio Tools for Universal Windows apps allow you to build a single universal app experience that can reach every device running Windows 10: phone, tablet, PC, and more. It includes the Microsoft Windows 10 Software Development Kit.

VisualStudio.Mac   1.0
Mac Extension for Visual Studio

Xamarin   4.12.3.77 (d15-9@e3f40b477)
Visual Studio extension to enable development for Xamarin.iOS and Xamarin.Android.

Xamarin Designer   4.16.13 (45a16efd4)
Visual Studio extension to enable Xamarin Designer tools in Visual Studio.

Xamarin Templates   1.1.128 (6f5ebb2)
Templates for building iOS, Android, and Windows apps with Xamarin and Xamarin.Forms.

Xamarin.Android SDK   9.1.4.2 (HEAD/8255f42fc)
Xamarin.Android Reference Assemblies and MSBuild support.

Xamarin.iOS and Xamarin.Mac SDK   12.2.1.12 (65ec520)
Xamarin.iOS and Xamarin.Mac Reference Assemblies and MSBuild support.
```

### Build Logs

[BuildLog.txt](https://github.com/xamarin/xamarin-macios/files/2727041/BuildLog.txt)

[LatestXamarinLogs.zip](https://github.com/xamarin/xamarin-macios/files/2727015/LatestXamarinLogs.zip)

### Example Project (If Possible)

> VS bug [#760827](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/760827)

**Comment by munkii** [comment]
I tried fixing installing the Alpha build referred to here, https://developercommunity.visualstudio.com/content/problem/406364/vsx1000-no-enough-information-has-been-provided-to.html but that did not fix it

**Comment by spouliot** [comment]
This code leading to this exception is executed on the windows-side only
```
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning : An error occurred getting directory checksum for App 
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning : Value cannot be null.
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning : Parameter name: path2
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning :    at System.IO.Path.Combine(String path1, String path2, String path3)
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning :    at Xamarin.Messaging.PlatformPath.GetServerBuildPath(String home, String appName, String sessionId, String path) in E:\A\_work\293\s\External\messaging\src\Xamarin.Messaging.Common\PlatformPath.cs:line 49
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning :    at Xamarin.Messaging.Ssh.ISshServiceExtensions.<GetDirectoryChecksumAsync>d__4.MoveNext() in E:\A\_work\293\s\External\messaging\src\Xamarin.Messaging.Ssh\Extensions\ISshServiceExtensions.cs:line 52
1>C:\Program Files (x86)\Microsoft Visual Studio\2017\Professional\MSBuild\Xamarin\iOS\Xamarin.Messaging.targets(82,3): warning : One or more errors occurred.
```

You might want to try directly on the mac (at least as a temporary workaround)

**Comment by spouliot** [comment]
@munkii glad using the Mac can work around your issue and thanks for confirming it!

**Comment by drewdz** [comment]
Exact same behaviour for me. Issues seemed to appear intermittently after updating VS on PC to 15.9.4. Using the Mac as a workaround is very tedious.

PC running VS Professional 2017 version 15.9.4 (config attached)
Mac build host is running macOS Mojave 10.14.2 and XCode 10.1

We have a rather large project for a client. At first it only affected another project in the suite but now it is affecting both iOS projects. I have compared project files, Info.plist, Entitlements.plist (etc) from before the problem and there are no significant differences.

Basically the VS parsing of the code completes, then the handover to the build host quietly fails. Watching ActivityMonitor on the Mac reveals that nothing gets started there.


[vs.txt](https://github.com/xamarin/xamarin-macios/files/2760563/vs.txt)

**Comment by Aleksey7151** [comment]
I'm having the same issue:
Launch failed. The app 'MyApp.Mobile.iOS' could not be launched on 'iSafonov'. Error: error MT1001: Could not find an application at the specified directory: /Users/aleksey/Library/Caches/Xamarin/mtbs/builds/MyApp.Mobile.iOS/e17314f941812c38a669a2524fc4c1ac/bin/iPhoneSimulator/Debug/MyApp.Mobile.iOS.app. Please check the logs for more details. 

Any updates?

**Comment by drewdz** [comment]
> I'm having the same issue:
> Launch failed. The app 'MyApp.Mobile.iOS' could not be launched on 'iSafonov'. Error: error MT1001: Could not find an application at the specified directory: /Users/aleksey/Library/Caches/Xamarin/mtbs/builds/MyApp.Mobile.iOS/e17314f941812c38a669a2524fc4c1ac/bin/iPhoneSimulator/Debug/MyApp.Mobile.iOS.app. Please check the logs for more details.
> 
> Any updates?

I know we're all holding our breath hoping someone can help. In the end the only thing that helped was to uninstall and reinstall VS. After reinstalling VS it worked fine for about 1.5 months; then suddenly one day it started again - same error.

I have no idea how to fix this.

**Comment by LeoJHarris** [comment]
Began hitting this issue too, why does Xamarin.iOS have to be sooooo painfull sometimes!

**Comment by emaf** [comment]
Hi all, 
Could you attach a **Diagnostic** build output after rebuilding your project (please make sure the app was not created on the Mac after rebuilding the project)? 
You can change the build output verbosity from `Tools -> Options -> Projects and Solutions -> Build and Run -> MSBuild project build output verbosity`

Also, it'd be great if you can share your env information (VS, macOS, Xcode versions) and confirm if this only happens for simulator builds and works for physical devices.

Thanks!

**Comment by NNebus** [comment]
Hi all, 

got the same issue. 
Havn't worked on my App for a few days (for a week or so). Meanwhile I updated my VS and macOS (Deployment-Server). 
I now wanted to make a test what I've done last and startet the iOS-Deployment on a real device and hit this error.

A few information:
iOS-Device: iPhone  6, iOS 12.4.3
Deployment-Server: Mac mini (2018), macOS Catalina (10.15)
Dev-System: Lenovo Thinkpad E580 (Lenovo 20KS001RGE), Windows 10.0.18362 (Windows 10 Pro, Build 18362), Visual Studio Community 2019, Version 16.3.7, .Net Framework 4.8.03752
Diagnostic-Log is attached. 

@emaf Do you need any other information?
[iOS-Logs-RYS.zip](https://github.com/xamarin/xamarin-macios/files/3802990/iOS-Logs-RYS.zip)

**Comment by drewdz** [comment]
@emaf Attached, is my Build output as well as the Debug output (zipped since the output was a bit on the large side :) )...
Hope you can shed some like here.

[debug.zip](https://github.com/xamarin/xamarin-macios/files/3803764/debug.zip)

**Comment by NNebus** [comment]
Okay guys, got it running again...
What i did: 

1. uninstall all dev-applications so the certificate will be removed from your device
2. create a native ios app with xcode
3. give the xcode-application the bundle-identifier of your xamarin-application
4. run the xcode-application to deploy it to the device (and having the cert re-installed)
5. start deployment from your windows maschine (make sure you have the same Bundle-Identifier here like in your xcode-app / -project!)
6. your xamarin app should now be able to run on your iOS-Device

Please follow this steps and give a feedback :)

**Comment by drewdz** [comment]
@NNebus Will give it a try and advise. Thanks :)

**Comment by emaf** [comment]
@drewdz @NNebus thanks for the logs you've attached!

@drewdz, in your case the build is being run offline even though you're effectively connected. There's another report tracking that issue and we already have a potential fix for it, if you'd like to get a build with that fix in place to confirm that you're facing that same problem please send me your email address so I can share it with you.

@NNebus based on your workaround, I don't think you're facing the same problem. But in any case, if you want me to confirm that, please attach **Diagnostic** build output after rebuilding your project. For that, you'd need to first change the output verbosity and then copy the content of the Build Output window to a text file. The logs you've share are just the Xamarin ones.

**Comment by drewdz** [comment]
@emaf Thanks so much for the info. Interesting, I followed (only step 1) of @NNebus 's work-around and that seemed have worked somehow. I simply removed the app from the device and did a claim/rebuild and redeploy and it worked.

It's great to hear there is a potential fix in the pipeline. As long as I am able to continue in this fashion I should be good to go. I'll see how long I can continue like this and let you know if it gets any worse. 

I'm wondering if the issues before were the same or just the same symptoms.

Again. Thanks so much for your assistance in this regard,

**Comment by pebr70** [comment]
Same problem with a solution: at first I restarted my mobile phone, my mac and my VisualStudio 2019 (on Windows 7). It didn't help.

Then I had clean the whole solution on VS, and deleted all bin and obj directories. Thereafter I had rebuild the solution, and deployed it. After that I started the app on VS (Win7) and he app ran properly. The problem was solved

**Comment by matteomonizza** [comment]
Same here. I've found that the build files are not in the folder that is specified in the output but are in: /Users/XXX/Library/Caches/Xamarin/mtbs/builds/MyApp.Mobile.iOS/e17314f941812c38a669a2524fc4c1ac/bin/iPhoneSimulator/Debug/**device-builds/ipad6.4-13.1.2**/MyApp.Mobile.iOS.app. So the app is in the wrong place!

**Comment by emaf** [comment]
Hi @pebr70 @matteomonizza, could you share the information I've asked for in https://github.com/xamarin/xamarin-macios/issues/5344#issuecomment-548942384. We've already identified a problem where VS is not building the project on the Mac even though it is effectively connected to it, with the requested information we can identify if that's the same root cause of the problem you're facing.

Thanks!

**Comment by matteomonizza** [comment]
Sorry, by now i've try the @pebr70 solution and finally it worked. Waiting for the problem to come again.!

**Comment by emaf** [comment]
Hi all,
The problem we found based on the attached logs were Visual Studio runs an offline build even though it's connected to a Mac si fixed in Visual Studio 2019 16.4 Preview 5 (https://visualstudio.microsoft.com/vs/preview/).

I'm closing this bug based on that fix, if you're still facing the MT1001 error please report a new issue since it's most likely to have a different root cause. 

Thanks!

**Comment by Aleksey7151** [comment]
Thanks a lot!

Отправлено с iPhone

**Comment by EvoPulseGaming** [comment]
Using 16.5.0 preview 1, and this issue is present  (using "Enable device-specific builds" option when this happened)

**Comment by DJKampouris** [comment]
I just ran into the same issue with VS 16.4.5 trying to deploy to a Physical device.
(Note : On the emulator it was working !)
I shortened Assembly name in csproj from 
`<AssemblyName>aaaa.bbb.Mobile.iOS</AssemblyName>`
to
`<AssemblyName>bbb.iOS</AssemblyName>`

and in info.plist

`<string>bbb</string>`

and the issue went away.

**Comment by Saturn** [comment]
> Same here. I've found that the build files are not in the folder that is specified in the output but are in: /Users/XXX/Library/Caches/Xamarin/mtbs/builds/MyApp.Mobile.iOS/e17314f941812c38a669a2524fc4c1ac/bin/iPhoneSimulator/Debug/**device-builds/ipad6.4-13.1.2**/MyApp.Mobile.iOS.app. So the app is in the wrong place!

I have found this too. I created a symlink pointing to the correct location and it now works.

**Comment by DanielGlos** [comment]
I just returned to one project after couple months of inactivity. I had to update XCode because my VS2019 version did not match. After the update I started getting this error saying that it can't find application in folder ...../device-builds/iphone7.2-12.4/ but the actual folder is called device-builds/iphone7.2-12.4.4/ so it seems like VS has some problems getting the path right. It's missing the last .4. If a rename this folder by removing the last .4 I'm able to run the application.

Any suggestions how to solve this?

**Comment by emaf** [comment]
Hi all,
The error mentioned in this issue is pretty generic since there could be different root causes for it. This was originally closed after fixing a connectivity problem, but it seems there're other problems related to the usage of device-specific builds, could you [report them from Visual Studio](https://docs.microsoft.com/visualstudio/ide/how-to-report-a-problem-with-visual-studio?view=vs-2019) so we can track and investigate them?

Btw, based on the latest comments a possible workaround could be to uncheck the `Enable device-specific builds` option from the **iOS Build** properties page, but please don't forget to report this problem from VS! 

Thanks for your help!

**Comment by DanielGlos** [comment]
@emaf I did reported it also from VS at the same time I reported it here. Also unchecking the device-specific builds works as a workaround and app launches properly

**Comment by emaf** [comment]
hey @DanielGlos, could you share the link to the report so I can take care of it?

Thanks!

**Comment by DanielGlos** [comment]
@emaf https://developercommunity.visualstudio.com/content/problem/952781/xamarinios-mt1001-error.html

**Comment by drungrin** [comment]
the isue is because of case sensitivity of osx filesystem. It is happening to me now.

It is searching for iPhone11.6-13.3

But on mac side, it is iphone11.6-13.3

**Comment by MichaelRumpler** [comment]
It's not (only) the casing.

On my Mac the folder is called "ipad6.4-13.3**.1**" but VS Win searches for "iPad6.4-13.3". It is also missing the ".1" at the end.
There are also [some posts](https://developercommunity.visualstudio.com/content/problem/952781/xamarinios-mt1001-error.html?childToView=958365#comment-958365) from users where that last part of the OS version is appended with a hyphen instead a dot like "iPhone8.1-13.3**-1**". But I never saw that.

**Comment by brepetti** [comment]
> It's not (only) the casing.
> 
> On my Mac the folder is called "ipad6.4-13.3.1" but VS Win searches for "iPad6.4-13.3". It is also missing the ".1" at the end.
> There are also [some posts](https://developercommunity.visualstudio.com/content/problem/952781/xamarinios-mt1001-error.html?childToView=958365#comment-958365) from users where that last part of the OS version is appended with a hyphen instead a dot like "iPhone8.1-13.3-1". But I never saw that.

Same problem. Folder name appears to be incorrect.  So much for working remote during a pandemic.

**Comment by KRA2008** [comment]
> 
> 
> It's not (only) the casing.
> 
> On my Mac the folder is called "ipad6.4-13.3.1" but VS Win searches for "iPad6.4-13.3". It is also missing the ".1" at the end.
> There are also [some posts](https://developercommunity.visualstudio.com/content/problem/952781/xamarinios-mt1001-error.html?childToView=958365#comment-958365) from users where that last part of the OS version is appended with a hyphen instead a dot like "iPhone8.1-13.3-1". But I never saw that.

I'm seeing the same thing. In my case it's doing "iPhone7.2-12.4" but the actual directory is "iPhone7.2-12.4.5"

**Comment by Dresel** [comment]
> 
> 
> > It's not (only) the casing.
> > On my Mac the folder is called "ipad6.4-13.3.1" but VS Win searches for "iPad6.4-13.3". It is also missing the ".1" at the end.
> > There are also [some posts](https://developercommunity.visualstudio.com/content/problem/952781/xamarinios-mt1001-error.html?childToView=958365#comment-958365) from users where that last part of the OS version is appended with a hyphen instead a dot like "iPhone8.1-13.3-1". But I never saw that.
> 
> Same problem. Folder name appears to be incorrect. So much for working remote during a pandemic.

Same here :/

**Comment by phickmanPDmB** [comment]
Had same issue after updating VS from 16.4.2 to 16.5.1.  Unchecking "Enable device-specific builds" aloud it to deploy to the device and debug.

**Comment by alexdecei** [comment]
Same for me. It might be the new vs update

**Comment by denisgeom** [comment]
Problem is still present with VS 2019 16.5.2 when using  Enable device-specific builds.
I can only debug and deploy App if Enable device-specific builds is unchecked.

**Comment by kyurkchyan** [comment]
Having the same issue VS 2019 16.5.2. If Enable device-specific builds is used app won't launched and fails with this exception

```
error MT1001: Could not find an application at the specified directory: /Users/****/Library/Caches/Xamarin/mtbs/builds/APP_NAME/0684320139e518760e75baec5b861669/bin/iPhone/Debug/device-builds/iPad4.4-11.4/APP_NAME.app
```

**Comment by alexdecei** [comment]
> Problem is still present with VS 2019 16.5.2 when using Enable device-specific builds.
> I can only debug and deploy App if Enable device-specific builds is unchecked.

Did not work for me !

**Comment by johnshardman** [comment]
Unchecking "Enable device-specific builds" allowed my app to deploy to physical devices.
Talk about wasting developer time. Using XF in 2020 feels like using XF v2 all over again :-(

**Comment by Saturn** [comment]
> Using XF in 2020 feels like using XF v2 all over again :-(

It seems to have come _quite_ a long way - But still a long way to go 😄 .

**Comment by jonebersole** [comment]
I realize this is a really old thread, but I had a different solution to this problem. I changed the AssemblyVersion property in the AssemblyInfo.cs file from 1.0.0.0 to 1.0.0.* so I could verify my revision numbers. It worked fine on my iPad Pro 4th edition and my iPhone 12 Max Pro, but kept failing with the above error on an iPad Air that I was testing with. I changed it back to 1.0.0.0 and it deploys to the iPad Air just fine. I then changed it back to 1.0.0.* and it still deploys to the iPad Air - go figure.
